### PR TITLE
Overlap compute and transfer more

### PR DESCRIPTION
**Pull request by anaruse**
This PR aims to overlap GPU kernels and data transfer more.

Current `ConcatWithAsyncTransfer` implementation uses global synchronization to prevent a GPU kernel read arrays that might be still being transferred to GPU. This implementation is safe but it blocks a CPU thread to launch GPU kernels until data transfer to GPU is finished.   With this PR, the global synchronized is removed and CUDA events are used instead to further overlap GPU kernels and data transfers w/o blocking a CPU thread to launch GPU kernels. You should specify CUDA stream used for compute kernels by an argument `compute_stream` like below, or global synchronization is used as before.

    converter = dataset.ConcatWithAsyncTransfer(compute_stream=chainer.backends.cuda.Stream.null)

### CODE REVIEW

File: chainer/dataset/convert.py [hunk 1]
```diff
@@ -205,13 +205,30 @@ class ConcatWithAsyncTransfer(object):
         stream (cupy.cuda.Stream): CUDA stream. If ``None``, a stream is
             automatically created on the first call. Data transfer operation
             is launched acynchrnously using the stream.
+        compute_stream(cupy.cuda.Stream): CUDA stream used for compute kernels.
+            If not ``None``, CUDA events are created/used to avoid global
+            synchronization and overlap execution of compute kernels and data
+            transfers as much as possible. If ``None``, global synchronization
+            is used instread.
```
Comment (mitmul) [inline chainer/dataset/convert.py]:
```suggestion
            is used instead.
```

File: chainer/dataset/convert.py [hunk 2]
```diff
@@ -205,13 +205,30 @@ class ConcatWithAsyncTransfer(object):
         stream (cupy.cuda.Stream): CUDA stream. If ``None``, a stream is
             automatically created on the first call. Data transfer operation
             is launched acynchrnously using the stream.
```
Comment (mitmul) [inline chainer/dataset/convert.py]:
```suggestion
            is launched asynchronously using the stream.
```

**Comment by anaruse** [comment]
Thanks, @mitmul . Typos were fixed.

File: chainer/dataset/convert.py [hunk 3]
```diff
@@ -239,6 +256,10 @@ def __call__(self, batch, device=None, padding=None):
         if device is not self._device:
             raise ValueError('device is different')
 
+        if self.compute_stream is not None:
+            self.event1.synchronize()
```
Comment (okuta) [inline chainer/dataset/convert.py]: Why do you use synchronize?

Comment (anaruse) [reply chainer/dataset/convert.py]: This synchronization is necessary to ensure that a CPU thread won't break the data of arrays that might be still being used by GPU kernels for previous iteration.

Suppose you are preparing input data for iteration _t_. Considering double-buffering scheme is used here, the array to be used for iteration _t_ was/is also used for iteration _t-2_. If you update/write the array for iteration _t_ while the array still being used by GPU kernels for iteration _t-2_, the results of iteration _t-2_ might get wrong.

How can we avoid this possible data corruption? The safe way is to make sure that all GPU kernels for iteration _t-2_ are finished, and the following codes are for that purpose.

    259        if self.compute_stream is not None:
    260            self.event1.synchronize()
    261            self.event1.record(stream=self.compute_stream)

As you can see, `self.event1.synchronize()` wait for the event that is "recorded" at previous iteration, more specifically, at the beginning of iteration _t-1_, in other word, the end of iteration _t-2_, on "compute stream". With this event synchronization, you can safely update the array.

File: chainer/dataset/convert.py [hunk 4]
```diff
@@ -345,19 +375,23 @@ def put(self, array):
...
-    def get(self):
+    def get(self, sync=True):
         """Returns the array of data transferred to a target device asynchronously.
 
-        This method first waits for completion of asynchrnous data trasfer
-        initiated by :meth:`put`, then returns the array on the target
-        device.
+        If sync is ``True``, the data of returned array is available in GPU
+        kernels. If sync is ``False``, the data of returned array might be
+        being transfered to GPU, so synchronizeaion must be done carefully by
```
Comment (grafi-tt) [inline chainer/dataset/convert.py]: synchronizeaion -> synchronization

Comment (anaruse) [reply chainer/dataset/convert.py]: Thanks!

**Comment by anaruse** [comment]
Thanks for your comment, @okuta .
I've just changed variable names based on your suggestion.

**Comment by okuta** [comment]
Thanks
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/dataset/convert.py (hunk 3 above)
Comment (okuta) [reply chainer/dataset/convert.py]: OK, I understood.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**